File: 이상욱/BOJ/step/BOJ1463.java
package boj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class BOJ1463 {
	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		int target = Integer.parseInt(br.readLine());

		int[] memo = new int[target + 1];
		Arrays.fill(memo, Integer.MAX_VALUE);
		memo[0] = 0;
		memo[1] = 0;

		for (int i = 2; i < memo.length; i++) {

			if (i % 3 == 0) {
				memo[i] = Math.min(memo[i], memo[i / 3] + 1);
			}

			if (i % 2 == 0) {
				memo[i] = Math.min(memo[i], memo[i / 2] + 1);
			}

			memo[i] = Math.min(memo[i], memo[i - 1] + 1);

		}
		System.out.println(memo[target]);
	}
}

File: 이상욱/BOJ/step/BOJ1463.kt
package level15

import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {

    val br = BufferedReader(InputStreamReader(System.`in`))

    val N = br.readLine().toInt()

    val dp = IntArray(N+1)

    dp.fill(Integer.MAX_VALUE)

    dp[1] = 0


    for (i in 2 until  dp.size){

        if(i % 3 ==0){
            dp[i] = (dp[i / 3] + 1).coerceAtMost(dp[i])
        }

        if(i % 2 ==0){
            dp[i] = (dp[i / 2] + 1).coerceAtMost(dp[i])
        }

        dp[i] = dp[i].coerceAtMost(dp[i - 1] + 1)

    }

    println(dp[N])


}